### PR TITLE
feat: Streaming predictions proof-of-concept

### PR DESCRIPTION
**Pull request by arkadyan**
#### Summary of changes
**Asana Ticket:** [Finish streaming POC](https://app.asana.com/0/555089885850811/1202995400172235)

- feat: Streaming predictions

Use a DynamicSupervisor for API prediction streams since they need to be
per route—a "firehose" ever everything is not supported by the API.
Start streams when a client subscribes to predictions for a route.
Register streams per route to prevent duplication.

Parse data into pure Prediction structs for the moment. We can think
harder about exactly what data format we want after demonstrating this
proof-of-concept.

- feat: Predictions Phoenix channel

- WIP(DO NOT MERGE): React predictions demo

**NB:** The first 2 commits are intended to be production-ready, but the 3rd is only a quick-and-dirty visual demonstration and is expected to be discarded/transformed.

---

Before getting review, please check the following:

* [ ] Does frontend functionality render and work correctly in IE?
* [ ] Have we load-tested any new pages or internal API endpoints that will receive significant traffic? [See load testing docs](./../apps/site/load_tests/README.md)
* [ ] Are interactive elements accessible to screen readers?
* [ ] Have you checked for tech debt you can address in the area you're working in?
* [ ] If this change involves routes, does it work correctly with pertinent "unusual" routes such as the combined Green Line, Silver Line, Foxboro commuter rail, and single-direction bus routes like the 170?
* [ ] Are the changes organized into self-contained commits with descriptive and well-formatted commit messages?

### CODE REVIEW

File: apps/predictions/lib/parser.ex [hunk 1]
```diff
@@ -13,30 +14,83 @@ defmodule Predictions.Parser do
           0 | 1,
           DateTime.t() | nil,
           non_neg_integer,
-          Prediction.schedule_relationship(),
+          Prediction.schedule_relationship() | nil,
```
Comment (thecristen) [inline apps/predictions/lib/parser.ex:17]: Good catch here.

**Review by thecristen — COMMENTED** [review]
Playing around with this, I notice we get predictions in the past (screenshot was taken 10:41am). Are old predictions not being replaces on later updates? Can we have the channel only output predictions that aren't in the past?

image

**Comment by arkadyan** [comment]
@thecristen Huh, I thought https://github.com/mbta/dotcom/pull/1421/files#diff-29f1b86ca05547f6e307f42931f9082aaadc38ea60095cbc3c33178069801323R46 was filtering out old predictions. I'll have to look more into that…

**Review by thecristen — COMMENTED** [review]
I noticed you added the departure and arrival times, which is cool. 👍🏼 
Other pieces that will be important for the display code come from the prediction's relationships. I noticed we have `trip_id` and `stop_id`, but in the context of a React component these are of very limited use - we can't fetch a trip `headsign` or `name`, or a stop `platform_code` without making additional requests to the backend. Any thoughts on how/where we should reconcile this?

**Review by thecristen — CHANGES_REQUESTED** [review]
Can we update this to have the predictions channel subscribe by both `routeId` _and_ `stopId`? After looking at how predictions are used, having channels defined by both stop+route makes more sense to me.
* Line diagram - each stop along a particular route
* Stop page - every route at a particular stop
* Transit Near Me - many routes + many stops

I'm also wondering if we can figure out why we're seeing predictions prior to `now`, but I can be convinced to leave that for a future iteration.

**Comment by thecristen** [comment]
Thanks Matt. Going to start a feature branch to iterate off this.

**Comment by thecristen** [comment]
Covered in #1476

**Pull request closed without merging**